pkg/webhook/bd_replacement.go

```diff
@@ -0,0 +1,409 @@
+	for i, oldPoolSpec := range commonPoolSpecs.oldSpec {
+		oldPoolSpec := oldPoolSpec
+		if ok, msg := bdr.IsReplacementValid(&oldPoolSpec, &commonPoolSpecs.newSpec[i]); !ok {
+			return false, msg
+		}
+	}
+	return true, ""
+}
+
+// getCommonPoolSpecs get the same pool specs from old persisted CSPC and the new CSPC after modification
+// which is not persisted yet.
+func getCommonPoolSpecs(cspcNew, cspcOld *apis.CStorPoolCluster) (*poolspecs, error) {
+	commonPoolSpecs := &poolspecs{
+		oldSpec: []apis.PoolSpec{},
+		newSpec: []apis.PoolSpec{},
+	}
+	for _, oldPool := range cspcOld.Spec.Pools {
+		oldNodeName, err := nodeselect.GetNodeFromLabelSelector(oldPool.NodeSelector)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, newPool := range cspcNew.Spec.Pools {
+			newNodeName, err := nodeselect.GetNodeFromLabelSelector(oldPool.NodeSelector)
+			if err != nil {
+				return nil, err
+			}
+			if oldNodeName == newNodeName {
+				commonPoolSpecs.oldSpec = append(commonPoolSpecs.oldSpec, oldPool)
+				commonPoolSpecs.newSpec = append(commonPoolSpecs.newSpec, newPool)
+				break
+			}
+		}
+	}
+	return commonPoolSpecs, nil
+}
+
+// IsReplacementValid validates the pool specs on CSPC for block device replacement case.
+func (bdr *BlockDeviceReplacement) IsReplacementValid(oldPoolSpec, newPoolSpec *apis.PoolSpec) (bool, string) {
+	newToOldBd := make(map[string]string)
+	for _, oldRg := range oldPoolSpec.RaidGroups {
+		oldRg := oldRg // pin it
+		if oldRg.Type == "" {
+			oldRg.Type = oldPoolSpec.PoolConfig.DefaultRaidGroupType
+		}
+		for _, newRg := range newPoolSpec.RaidGroups {
+			newRg := newRg // pin it
+			if IsRaidGroupCommon(oldRg, newRg) {
+				if IsBlockDeviceReplacementCase(&oldRg, &newRg) {
+					if ok, msg := bdr.IsBDReplacementValid(&newRg, &oldRg); !ok {
+						return false, msg
+					}
+					newBD := GetNewBDFromRaidGroups(&newRg, &oldRg)
+					for k, v := range newBD {
+						newToOldBd[k] = v
+					}
+				}
+				break
+			}
+		}
+	}
+
+	for newBD, oldBD := range newToOldBd {
+		err := createBDC(newBD, oldBD, bdr.OldCSPC)
+		if err != nil {
+			return false, err.Error()
+		}
+	}
+	return true, ""
+}
+
+// IsRaidGroupCommon returns true if the provided raid groups are the same raid groups.
+func IsRaidGroupCommon(rgOld, rgNew apis.RaidGroup) bool {
+	oldBdMap := make(map[string]bool)
+	for _, bd := range rgOld.BlockDevices {
+		oldBdMap[bd.BlockDeviceName] = true
+	}
+
+	for _, bd := range rgNew.BlockDevices {
+		if oldBdMap[bd.BlockDeviceName] {
+			return true
+		}
+	}
+	return false
+}
+
+// IsBlockDeviceReplacementCase returns true if the edit/update of CSPC can trigger a blockdevice
+// replacement.
+func IsBlockDeviceReplacementCase(newRaidGroup, oldRaidGroup *apis.RaidGroup) bool {
+	count := GetNumberOfDiskReplaced(newRaidGroup, oldRaidGroup)
+	return count >= 1
+}
+
+// GetNumberOfDiskReplaced returns the nuber of disk replaced in raid group.
+func GetNumberOfDiskReplaced(newRG, oldRG *apis.RaidGroup) int {
```

Comment:
we have the similar func has been added as part of other PR for disk replacement feature, we can refactor these in subsequent PR's

Reply:
Sure. Will take it in a later PR.